Add ITreeAlpha and related APIs (#24225)
microsoft/FluidFramework: .changeset/tired-places-cough.md

---
"fluid-framework": minor
"@fluidframework/tree": minor
---
---

> "section": feature

Reviewer: Feature is technically correct but I think we're still putting all SharedTree changes in their own section?
```suggestion
"section": tree
```

Author: yes, this was a mistake.

---

> Provide access to tree content and stored schema without requiring a compatible view schema

Author: ```suggestion
Provide alpha APIs for accessing tree content and stored schema without requiring a compatible view schema
```

> Provide access to tree content and stored schema without requiring a compatible view schema

Reviewer: Is this still accurate? My main nit about the existing version if that a compatible view schema seems to still be required to access the data in particular, but now we can generate it at runtime.
```suggestion
Provide alpha APIs for accessing tree content and stored schema when the view schema isn't known at compile time
```

Author: This PR exposes ITreeAlpha, which you can get without a view schema, that has:
```typescript
 exportSimpleSchema(): SimpleTreeSchema;
exportVerbose(): VerboseTree | undefined;
```

Which allow access to the schema and content respectively.

It is important to note that these can be useful if even if you have a view schema but it isn't compatible with the document. Thus I think my original phrasing is more relevant than the proposed change here.

Author: I have applied part of this suggestion

> Adds `ITreeAlpha` (Which ITree can be down casted to) that exposes access to both the tree content and the schema.
> This allows inspecting the content saved in a shared tree in a generic way that can work on any shared tree.

Reviewer: ```suggestion
Adds a `ITreeAlpha` interface (which `ITree` can be down-casted to) that provides access to both the tree content and the schema.
This allows inspecting the content saved in a SharedTree in a generic way that can work on any SharedTree.
```

> This can be combined with the existing `generateSchemaFromSimpleSchema` to generate a schema that can be used with [`IIree.viewWith`](https://fluidframework.com/docs/api/fluid-framework/viewabletree-interface#viewwith-methodsignature) to allow constructing a [`TreeView`](https://fluidframework.com/docs/api/fluid-framework/treeview-interface) for any shared tree, regardless of its schema.

Reviewer: ```suggestion
This can be combined with the existing `generateSchemaFromSimpleSchema` to generate a schema that can be used with [`IIree.viewWith`](https://fluidframework.com/docs/api/fluid-framework/viewabletree-interface#viewwith-methodsignature) to allow constructing a [`TreeView`](https://fluidframework.com/docs/api/fluid-framework/treeview-interface) for any SharedTree, regardless of its schema.
```

> Note that the resulting TypeScript typing for such a view will not be friendly: the TreeView APIs are designed for statically known schema. With care and a lot of type casts use is possible, but not recommended when avoidable: see disclaimer on `generateSchemaFromSimpleSchema`.

Reviewer: ```suggestion
Note that the resulting TypeScript typing for such a view will not be friendly: the `TreeView` APIs are designed for statically known schema. Using them is possible with care and a lot of type casts but not recommended if it can be avoided: see disclaimer on `generateSchemaFromSimpleSchema`.
```

---
"fluid-framework": minor
"@fluidframework/tree": minor
---
---
"section": feature
---

Provide access to tree content and stored schema without requiring a compatible view schema

> Adds `ITreeAlpha` (Which ITree can be down casted to) that exposes access to both the tree content and the schema.

Author: TreeView has some nuanced complications with type parameters which make casting it correctly harder than this case.

I think this is fine without such a function, though we could add one later if people want one.

> Adds `ITreeAlpha` (Which ITree can be down casted to) that exposes access to both the tree content and the schema.

Reviewer 2: Actually, since we have followed this pattern consistently of having Foo and FooAlpha types, it might make sense to have a single top-level `asAlpha` API with various overloads. For example:

```ts
asAlpha(view: TreeView): TreeViewAlpha;
asAlpha(tree: ITree): ITreeAlpha;
// ...
asAlpha(value: TreeView | ITree): TreeViewAlpha | ITreeAlpha {
    return value;
}

> Adds `ITreeAlpha` (Which ITree can be down casted to) that exposes access to both the tree content and the schema.

Author: Thats a neat idea to avoid cluttering the namespace. I like it (though not as part of this change).
